File: Hot 100/Backtracing/0078_Subsets.java
//Backtracing
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();

        backtrack(nums, 0, new ArrayList<>(), ans);

        return ans;
    }

    private void backtrack(int[] nums, int idx, List<Integer> path, List<List<Integer>> ans) {
        if (idx == nums.length) {
            ans.add(new ArrayList<>(path));
            return;
        }

        backtrack(nums, idx + 1, path, ans);

        path.add(nums[idx]);
        backtrack(nums, idx + 1, path, ans);
        path.remove(path.size() - 1);
    }
}

File: Hot 100/Backtracing/0078_subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
      ans = []
      path = []

      def dfs(idx: int):
        if idx == len(nums):
          ans.append(path.copy())
          return

        dfs(idx + 1)

        path.append(nums[idx])
        dfs(idx + 1)
        path.pop()

      dfs(0)
      return ans
